src/tensor/aggregate.nim

```diff
@@ -14,26 +14,98 @@
 
 import  ./data_structure,
         ./higher_order,
-        ./operators_blas_l1
+        ./operators_blas_l1,
+        ./math_functions,
+        math
 
 # ### Standard aggregate functions
 # TODO consider using stats from Nim standard lib: https://nim-lang.org/docs/stats.html#standardDeviation,RunningStat
 
-proc sum*[T: SomeNumber](t: Tensor[T]): T =
-  ## Compute the sum of all elements of T
-
-  t.reduceT():
+proc sum*[T](t: Tensor[T]): T =
+  ## Compute the sum of all elements
+  t.reduce_inline():
     x+=y
 
-proc sum*[T: SomeNumber](t: Tensor[T], axis: int): Tensor[T] {.noInit,noInit.} =
-  ## Compute the sum of all elements of T along an axis
-  t.reduceAxisT(axis):
+proc sum*[T](t: Tensor[T], axis: int): Tensor[T] {.noInit,noInit.} =
+  ## Compute the sum of all elements along an axis
+  t.reduce_axis_inline(axis):
     x+=y
 
+proc mean*[T: SomeInteger](t: Tensor[T]): T {.inline.}=
+  ## Compute the mean of all elements
+  t.sum div t.size.T
+
+proc mean*[T: SomeInteger](t: Tensor[T], axis: int): Tensor[T] {.noInit,inline.}=
+  ## Compute the mean along an axis
+  t.sum(axis) div t.shape[axis].T
+
 proc mean*[T: SomeReal](t: Tensor[T]): T {.inline.}=
-  ## Compute the mean of all elements of T
+  ## Compute the mean of all elements
   t.sum / t.size.T
 
 proc mean*[T: SomeReal](t: Tensor[T], axis: int): Tensor[T] {.noInit,inline.}=
-  ## Compute the mean of T along an axis
+  ## Compute the mean along an axis
   t.sum(axis) / t.shape[axis].T
+
+proc min*[T](t: Tensor[T]): T =
+  ## Compute the min of all elements
+  t.reduce_inline():
+    x = min(x,y)
+
+proc min*[T](t: Tensor[T], axis: int): Tensor[T] {.noInit.} =
+  ## Compute the min along an axis
+  t.reduce_axis_inline(axis):
+    for ex, ey in mzip(x,y):
+      ex = min(ex,ey)
+
+proc max*[T](t: Tensor[T]): T =
+  ## Compute the max of all elements
+  t.reduce_inline():
+    x = max(x,y)
+
+proc max*[T](t: Tensor[T], axis: int): Tensor[T] {.noInit.} =
+  ## Compute the max along an axis
+  t.reduce_axis_inline(axis):
+    for ex, ey in mzip(x,y):
+      ex = max(ex,ey)
+
+proc variance*[T: SomeReal](t: Tensor[T]): T =
+  ## Compute the variance of all elements
+  ## The normalization is by the (n-1), like in the formal definition
+  let mean = t.mean()
+  result = t.fold_inline() do:
+    # Initialize to the first element
+    x = square(y - mean)
+  do:
+    # Fold in parallel by summing remaning elements
+    x += square(y - mean)
+  do:
+    # Merge parallel folds
+    x += y
+  result /= (t.size-1).T
+
+proc variance*[T: SomeReal](t: Tensor[T], axis: int): Tensor[T] {.noInit.} =
+  ## Compute the variance of all elements
+  ## The normalization is by the (n-1), like in the formal definition
```

Comment:
I would say "The normalization is by (n-1), also known as Bessel's correction, which partially correct the bias of estimating a population variance from a sample of this population"